File: jenjin-core/src/main/java/com/binarymonks/jj/pools/Pool.java
package com.binarymonks.jj.pools;

import com.badlogic.gdx.utils.Array;

public class Pool<POOLED> {

    PoolManager<POOLED> manager;
    Array<POOLED> pool = new Array<>();

    public Pool(PoolManager<POOLED> manager) {
        this.manager = manager;
    }

    public POOLED getNew() {
        if (pool.size == 0) {
            return manager.create_new();
        }
        return pool.pop();
    }

    public void add(POOLED pooled) {
        manager.reset(pooled);
        pool.add(pooled);
    }

    public void clear() {
        for (POOLED pooled : pool) {
            manager.dispose(pooled);
        }
        pool.clear();
    }
}

File: jenjin-core/src/main/kotlin/com/binarymonks/jj/core/pools/Pool.kt
package com.binarymonks.jj.core.pools

import com.badlogic.gdx.utils.Array

class Pool<POOLED>(internal var manager: PoolManager<POOLED>) {
    internal var pool = Array<POOLED>()

    val new: POOLED
        get() {
            if (pool.size == 0) {
                return manager.create_new()
            }
            return pool.pop()
        }

    fun add(pooled: POOLED) {
        manager.reset(pooled)
        pool.add(pooled)
    }

    fun clear() {
        for (pooled in pool) {
            manager.dispose(pooled)
        }
        pool.clear()
    }
}
